projects/feature-activation/0005-feature-activation-for-transactions.md

```diff
@@ -0,0 +1,359 @@
+
+## Overview
+
+### Premises
+
+1. After an `ACTIVE` block, all next blocks will also be `ACTIVE` (considering the same blockchain).
+2. Feature states for transactions are only `ACTIVE` or `INACTIVE`.
+
+### Requirements
+
+Using the premises above, we must define a function that returns a state, given a transaction (note: we actually return multiple states for multiple features). That function must satisfy the following requirements:
+
+1. All transactions that are received after (time-wise) an `ACTIVE` block in the best chain, must also be `ACTIVE`.
+2. When all transactions are ordered by timestamp, there must not be an `INACTIVE` transaction after an `ACTIVE` transaction. In other words, as soon as one transaction becomes `ACTIVE`, every future transaction will also be `ACTIVE`.
+
+### Definitions
+
+Repeated here from the Guide-level section.
+
+#### State of a transaction
+
+For a feature, a transaction is considered `ACTIVE` if its timestamp is greater than one Evaluation Interval after the _first_ `ACTIVE` block's timestamp.
+
+#### Dealing with reorgs
+
+If we've passed one Evaluation Interval after the _first_ `ACTIVE` boundary block AND the reorg includes that _first_ `ACTIVE` boundary block, the reorg is considered invalid and the full node halts execution.
+
+## Retrieving Feature States for Transactions
+
+Analogously to the Feature Activation for Blocks, the state for a transaction will be retrieved from a method in the `FeatureService`. Considering the definition above, a pseudo reference-implementation is provided:
+
+```python
+class FeatureService:
+    def is_feature_active_for_transaction(self, *, transaction: Transaction, feature: Feature) -> bool:
+        first_active_block = self._get_first_active_block(feature)
+
+        if not first_active_block:
+            return False
+
+        # equivalent to two weeks
+        avg_time_between_boundaries = self._feature_settings.evaluation_interval * settings.AVG_TIME_BETWEEN_BLOCKS
+        activation_threshold = first_active_block.timestamp + avg_time_between_boundaries
+        # We also use the MAX_FUTURE_TIMESTAMP_ALLOWED to take into account that we can receive a tx from the future
+        is_active = transaction.timestamp > activation_threshold + settings.MAX_FUTURE_TIMESTAMP_ALLOWED
+
+        return is_active
+
+    def _get_first_active_block(self, feature: Feature) -> Optional[Block]:
+      """
+      Return the first ever block that became ACTIVE for a specific feature (always a boundary block),
+      or None if this feature is not ACTIVE.
+
+      This can be implemented by recursively hopping boundary blocks until
+      we find a block that is ACTIVE and has a parent that is LOCKED_IN.
+      """
+      raise NotImplementedError
+```
+
+## Dealing with reorgs
+
+Considering the definition above, a function must be called in the `ConsensusAlgorithm.update()` method that determines whether a reorg is invalid. If it is, the full node will **halt execution**.
```

Comment:
What exactly "halts execution" mean here? Would the fullnode's main process just exit with a non-zero code? What would we do with the database? Leave it in a non-consistent state? Wouldn't it be better to revert the re-org back to where it was before if a re-org is invalid?

Reply:
> What exactly "halts execution" mean here? Would the fullnode's main process just exit with a non-zero code?

Exactly. It's probably a good idea to add this to the text.

> What would we do with the database? Leave it in a non-consistent state?

Yes, it would work just like when we have a fail during loading, requiring the database to be reset. That's what I call "manual intervention".

> Wouldn't it be better to revert the re-org back to where it was before if a re-org is invalid?

Probably, but that would require us to implement the "database commit and rollback" mechanism in the storage which we currently don't have. I agree it would be nice, but I think that's a separate project since it touches critical code. The loading case, mentioned above, could also benefit from this.

But even if we do that, wouldn't the block that caused the reorg eventually be received again? Or you mean, reverting the reorg _and then_ halting the full node? I don't know if there would be many advantages in doing that, as the full node wouldn't be able to sync again anyway.

Also, we have to consider the likelihood of this ever happening, described in the [Likelihood of halting the full node](https://github.com/HathorNetwork/rfcs/blob/feat/feature-activation/transactions/projects/feature-activation/0005-feature-activation-for-transactions.md#likelihood-of-halting-the-full-node) section.

Reply:
Got it. I think it makes sense, if I have any considerations about the "likelihood" (because it might make sense to consider if it's possible for it to happen accidentally or the effort required to do it maliciously) I'll add a comment to that section.